prov/efa/src/rxr/rxr_pkt_cmd.c

```diff
@@ -814,6 +814,83 @@ fi_addr_t rxr_pkt_insert_addr(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry
 	return rdm_addr;
 }
 
+static inline
+void rxr_pkt_handle_zero_copy_recv_completion(struct rxr_ep *ep,
+					      struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_rx_entry *rx_entry;
+	struct rxr_base_hdr *base_hdr;
+	size_t expected_hdr_size, received_hdr_size;
+	/*
+	 * in zero copy receive mode, EFA device must be the
+	 * only source of incoming packets. So shm communication
+	 * cannot be used, therefore peer must NOT be local
+	 */
+	assert(!rxr_ep_get_peer(ep, pkt_entry->addr)->is_local);
+	ep->posted_bufs_efa--;
+
+	base_hdr = rxr_get_base_hdr(pkt_entry->pkt);
+	if (base_hdr->type != RXR_EAGER_MSGRTM_PKT) {
+		FI_WARN(&rxr_prov, FI_LOG_CQ,
+			"invalid pkt type %d for zero copy receive mode\n",
+			rxr_get_base_hdr(pkt_entry->pkt)->type);
+		efa_eq_write_error(&ep->util_ep, FI_EIO, -FI_EIO);
+		return;
+	}
+
+	/* In zero copy receive mode, we will not send handshake packet
+	 * back to peer. This is to ensure all packets we received have
+	 * the same header size.
+	 */
+
+
+	/*
+	 * In zero copy receive mode, an RX entry was created in rxr_msg_generic_recv(),
+	 * the rx_entry was put in ep->rx_list, and ready to be matched.
+	 * At the same time, rxr_msg_generic_recv() posted a packet entry to receive
+	 * data, which is the one we are processing.
+	 * Calling rxr_pkt_get_msgrtm_rx_entry() should match that RX entry with this
+	 * paket entry, and return us a matched RX entry.
+	 */
```

Comment:
Can be placed in one comment block.

Reply:
Since they are taking about two things, I prefer keep them separated.

Reply:
at a minimum, don't have two lines between the comment blocks :)

Reply:
Fixed in new revision.